### ISSUE TITLE
Compatibility with magic-latex-buffer

### ISSUE THREAD

**Issue by antoine-levitt** [dandavison/xenops#9]
Hi and thanks for this package, it looks nice! I was trying to use it but ran into several issues. My latex configuration is a mess so it might not be the only issue, but I think the main blocker is compatibility with the (awesome) https://github.com/zk-phi/magic-latex-buffer. Could this be fixed?

**Comment by dandavison** [comment]
Hi @antoine-levitt, thanks. I've installed `magic-latex-buffer` (thanks for the recommendation!) and am using it together with `xenops`. I haven't noticed any problems yet (in fact I think I like it),  but perhaps that's because I'm new to magit-latex-buffer. Can you tell me what problems you found? Here's an example of what it looks like for me. Is this how it should look?

<table><tr><td><img width=500px src="https://user-images.githubusercontent.com/52205/98445748-62ad7080-20e7-11eb-8ed3-bf18fd85d6f1.png" alt="image" /></td></tr></table>

**Comment by dandavison** [comment]
OK I am hitting some bugs related to cursor movement and failure to trigger reveal/render when the cursor enters/exits a block. For example below, the cursor is returning to the beginning-of-line inappropriately. Is this the sort of thing you encountered? (These seem like they should be fixable.)

![xenops-3-1](https://user-images.githubusercontent.com/52205/98446820-5e388600-20ee-11eb-8f26-afbc6a69358f.gif)

**Comment by antoine-levitt** [comment]
Even the basic things failed to work for me. I'll clean up my latex setup a bit and see if I can get something reproducible.

**Comment by dandavison** [comment]
OK. I'm happy to help troubleshoot here. Have you tried the `M-x xenops-doctor` command? That should detect problems with your setup and suggest resolutions. This is obvious but probably makes sense to try first without `magic-latex-buffer`.

**Comment by antoine-levitt** [comment]
xenops-doctor is healthy. xenops-dwim does render the inline math ($$) OK (though very small, but I'll try to customize that), but the equation blocks are wrong. I get
```
redisplay--pre-redisplay-functions: (wrong-type-argument listp #("▽" 0 1 (display nil))) [546 times]
```
which I think occurs because magic-latex renders \begin as ▽. You should be able to reproduce with any begin/end block.

**Comment by antoine-levitt** [comment]
Hm apparently not according to your video. Here's the backtrace, if that tells you anything
```
Debugger entered--Lisp error: (wrong-type-argument listp #("▽" 0 1 (display nil)))
  xenops-util-parse-image-at(1031)
  xenops-element-get-image((:type block-math :begin 1031 :begin-content 1045 :end-content 1073 :end 1085 :delimiters ("^[ \11]*\\\\begin{\\(align\\|equation\\|tikzpicture\\|gath..." "^[ \11]*\\\\end{\\(align\\|equation\\|tikzpicture\\|gather...")))
  xenops-math-handle-element-transgression(#<window 3 on main.tex<2020-09-conformal-integration>> 1037 left)
  cursor-sensor--detect(#<window 3 on main.tex<2020-09-conformal-integration>>)
  run-hook-with-args(cursor-sensor--detect #<window 3 on main.tex<2020-09-conformal-integration>>)
  redisplay--pre-redisplay-functions(t)
  apply(redisplay--pre-redisplay-functions t)
  #f(advice-wrapper :before ignore redisplay--pre-redisplay-functions)(t)
  redisplay_internal\ \(C\ function\)()
```

**Comment by antoine-levitt** [comment]
Even without magic-latex-buffer, when doing xenops-dwim outside of any equation I get little warning signs saying it's missing a begin{document}. I think that's because I did $a+b$ to try xenops but I did that before the begin{document}, and now it's stuck with this $a+b$ (that is not in the tex anymore). I guess that's a separate bug though.

**Comment by dandavison** [comment]
> render the inline math ($$) OK (though very small, but I'll try to customize that),

Yes. Try `(setq xenops-math-image-scale-factor 1.0)`. I'm using that personally. I think perhaps the default is too small, although the right value may vary according to user's display resolution etc (what's your OS and display resolution out of interest? Do the SVGs render crisply for you?)

> when doing xenops-dwim outside of any equation I get little warning signs saying it's missing a begin{document}. I think that's because I did $a+b$ to try xenops but I did that before the begin{document}, and now it's stuck with this $a+b$ (that is not in the tex anymore).

Thanks! I just reproduced that. You can fix it with this command: `M-x xenops-clear-latex-preamble-cache` I'll try to find a proper fix for it; maybe xenops should not be caching a preamble that does not compile successfully or something like that.

**Comment by dandavison** [comment]
> redisplay--pre-redisplay-functions: (wrong-type-argument listp #("▽" 0 1 (display nil))) [546 times]

Yes, I can reproduce that also. I'll try to get it fixed today or tomorrow.

Thanks for all this! I'm happy to help with any more troubleshooting or configuring.

**Comment by antoine-levitt** [comment]
> (what's your OS and display resolution out of interest? Do the SVGs render crisply for you?)

It's 1920x1080. SVGs are pretty nice!

> I'm using that personally. I think perhaps the default is too small, although the right value may vary according to user's display resolution etc

Any way to set that based on the font size so that the fonts match?

> Try (setq xenops-math-image-scale-factor 1.0). 

Even with that the images are pretty small. Here's how it looks like: 
![Screenshot at 2020-11-07 21-20-30](https://user-images.githubusercontent.com/88065/98450723-4f2df400-213f-11eb-84d1-75649f301565.png)

**Comment by antoine-levitt** [comment]
> You can fix it with this command: M-x xenops-clear-latex-preamble-cache

Thanks!

That made it work. Now I can't get in the equations with my usual C-m, but RET does work. xenops-reveal-on-entry doesn't seem to though... I'll play with it a bit.

**Comment by antoine-levitt** [comment]
I'm on linux btw, and my frame font is Noto Mono 16.

**Comment by dandavison** [comment]
> redisplay--pre-redisplay-functions: (wrong-type-argument listp #("▽" 0 1 (display nil))) [546 times]

I've fixed this in master (f6be2308452fa955d3b5811319e22bbc45916abd). Let me know if there are any other bugs with `magic-latex-mode`, but I'm hoping it's now usable.

>> Try (setq xenops-math-image-scale-factor 1.0).

> Even with that the images are pretty small.

Sorry, I should have said, the old (small) images will be cached. So after changing something like `xenops-math-image-scale-factor`, the simplest way to get rid of all the old small images is to kill the buffer, delete the files in the `xenops-cache-directory`, and open the file again. Alternatively, a double prefix argument `C-u C-u M-x xenops-dwim` regenerates, ignoring the cached version.

You should be able to get them the size you want by experimenting with the value of `xenops-math-image-scale-factor` and deleting the old cache files, right?

> Any way to set that based on the font size so that the fonts match?

If it's possible to do that calculation that sounds like a great idea.

**Comment by antoine-levitt** [comment]
> Nothing automated currently but if it's possible to do that calculation that sounds like a great idea.

(default-line-height) is the height of the line in pixels, is that enough information?

**Comment by antoine-levitt** [comment]
OK I was able to get it to work. Scale factor of 1.7 looks about right for me. There are other issues but I will open separate issues for them.
